GitUI/CommandsDialogs/SettingsDialog/ShellExtension/ShellExtensionManager.cs

```diff
@@ -0,0 +1,115 @@
+        /// <summary>
+        ///  Checks if the shell extension files are found at the known location.
+        /// </summary>
+        /// <returns><see langword="true"/> the extension files are found at the known location; otherwise <see langword="false"/>.</returns>
+        public static bool FilesExist()
+        {
+            string path32 = FindFileInBinFolders(CommonLogic.GitExtensionsShellEx32Name);
+            string path64 = FindFileInBinFolders(CommonLogic.GitExtensionsShellEx64Name);
+            return !(string.IsNullOrEmpty(path32) || (Environment.Is64BitOperatingSystem && string.IsNullOrEmpty(path64)));
+        }
+
+        /// <summary>
+        ///  Checks if the shell extension is properly registered in the system.
+        /// </summary>
+        /// <returns><see langword="true"/> if the extension is registered; otherwise <see langword="false"/>.</returns>
+        public static bool IsRegistered()
+        {
+            return !string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.LocalMachine, @"Software\Microsoft\Windows\CurrentVersion\Shell Extensions\Approved", "{3C16B20A-BA16-4156-916F-0A375ECFFE24}"))
+                && !string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot, @"*\shellex\ContextMenuHandlers\GitExtensions2"))
+                && !string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot, @"Directory\shellex\ContextMenuHandlers\GitExtensions2"))
+                && !string.IsNullOrEmpty(CommonLogic.GetRegistryValue(Registry.ClassesRoot, @"Directory\Background\shellex\ContextMenuHandlers\GitExtensions2"));
+        }
+
+        /// <summary>
+        /// Register shell extensions
+        /// </summary>
+        /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
+        /// <exception cref="Win32Exception">If user canceled elevation dialog (when ex.NativeErrorCode == 1223)</exception>
+        /// <exception cref="Exception">Other potential error</exception>
+        public static void Register() => RunRegSvrForShellExtensionDlls("{0}");
+
+        /// <summary>
+        /// Unregister shell extensions
+        /// </summary>
+        /// <exception cref="FileNotFoundException">If at least one necessary for registration file wasn't found</exception>
+        /// <exception cref="Win32Exception">If user canceled elevation dialog (when ex.NativeErrorCode == 1223)</exception>
+        /// <exception cref="Exception">Other potential error</exception>
+        public static void Unregister() => RunRegSvrForShellExtensionDlls("/u {0}");
+
+        private static void RunRegSvrForShellExtensionDlls(string argumentsPattern)
+        {
+            if (Environment.Is64BitOperatingSystem)
+            {
+                RunRegSvrForSingleDll(CommonLogic.GitExtensionsShellEx64Name, argumentsPattern);
+            }
+
+            RunRegSvrForSingleDll(CommonLogic.GitExtensionsShellEx32Name, argumentsPattern);
```

Comment:
I'd wrap this in try...catch and throw UseExternal exception. This will allow us not to catch exceptions at callsites.